src/proxy/outbound.rs

```diff
@@ -380,7 +381,69 @@ impl OutboundConnection {
             None => return Err(Error::UnknownSource(downstream)),
         };
 
+        // to Service with waypoint
+        if let Some(target_address) = self
+            .pi
+            .state
+            .fetch_destination(&Destination::Address(NetworkAddress {
+                network: self.pi.cfg.network.clone(),
+                address: target.ip(),
+            }))
+            .await
+        {
+            match target_address {
+                crate::state::workload::address::Address::Workload(_w) => {
+                    // this is to workload, fall through
+                }
+                crate::state::workload::address::Address::Service(s) => {
+                    if let Some(wp) = s.waypoint.clone() {
+                        let waypoint_vip = match wp.destination {
+                            Destination::Address(a) => a.address,
+                            Destination::Hostname(_) => todo!(), // it has a waypoint with an invalid addy... error?
+                        };
+                        let waypoint_vip = SocketAddr::new(waypoint_vip, wp.hbone_mtls_port);
+                        let waypoint_us = self
+                            .pi
+                            .state
+                            .fetch_upstream(&self.pi.cfg.network, waypoint_vip)
+                            .await
+                            .unwrap(); // ugh, TODO NOT OK
+
+                        let waypoint_workload = waypoint_us.workload;
+                        let waypoint_ip = self
+                            .pi
+                            .state
+                            .load_balance(
+                                &waypoint_workload,
+                                &source_workload,
+                                self.pi.metrics.clone(),
+                            )
+                            .await
+                            .unwrap(); // ugh, TODO NOT OK
+
+                        let waypoint_socket_address =
+                            SocketAddr::new(waypoint_ip, waypoint_us.port);
+                        let destination_service = ServiceDescription::try_from(&*s).ok();
+
+                        return Ok(Request {
```

Comment:
can we abstract a fine grained function to reuse

Reply:
Do you mean some sort of "request builder" instead of returning literals all the time?

Reply:
yes

Reply:
I'd rather not address in this PR if it's all the same. I opened #846 to track work on this idea.